File: promptdev-backend/src/main/java/com/promptdev/service/JiraService.java
package com.promptdev.service;

import com.promptdev.dto.jira.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.Map;

/**
 * Service for Jira Server REST API integration.
 * Only available when jira.base-url is configured.
 */
@Service
@ConditionalOnBean(name = "jiraRestClient")
public class JiraService {

    private static final Logger log = LoggerFactory.getLogger(JiraService.class);

    private final RestClient jiraRestClient;

    public JiraService(RestClient jiraRestClient) {
        this.jiraRestClient = jiraRestClient;
    }

    /**
     * Search issues using JQL
     */
    public JiraSearchResponse searchIssues(String jql, int startAt, int maxResults) {
        log.info("Searching Jira issues with JQL: {}", jql);
        return jiraRestClient.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/search")
                        .queryParam("jql", jql)
                        .queryParam("startAt", startAt)
                        .queryParam("maxResults", maxResults)
                        .build())
                .retrieve()
                .body(JiraSearchResponse.class);
    }

    /**
     * Get a single issue by key
     */
    public JiraIssueResponse getIssue(String issueKey) {
        log.info("Fetching Jira issue: {}", issueKey);
        return jiraRestClient.get()
                .uri("/issue/{issueKey}", issueKey)
                .retrieve()
                .body(JiraIssueResponse.class);
    }

    /**
     * Get available transitions for an issue
     */
    public JiraTransitionResponse getTransitions(String issueKey) {
        log.info("Fetching transitions for issue: {}", issueKey);
        return jiraRestClient.get()
                .uri("/issue/{issueKey}/transitions", issueKey)
                .retrieve()
                .body(JiraTransitionResponse.class);
    }

    /**
     * Transition an issue to a new status
     */
    public void transitionIssue(String issueKey, String transitionId) {
        log.info("Transitioning issue {} with transition {}", issueKey, transitionId);
        JiraTransitionRequest request = new JiraTransitionRequest(
                new JiraTransitionRequest.JiraTransitionInfo(transitionId));

        jiraRestClient.post()
                .uri("/issue/{issueKey}/transitions", issueKey)
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .toBodilessEntity();
    }

    /**
     * Add a comment to an issue
     */
    public void addComment(String issueKey, String comment) {
        log.info("Adding comment to issue: {}", issueKey);
        JiraCommentRequest request = new JiraCommentRequest(comment);

        jiraRestClient.post()
                .uri("/issue/{issueKey}/comment", issueKey)
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .toBodilessEntity();
    }

    /**
     * Assign an issue to a user
     */
    public void assignIssue(String issueKey, String username) {
        log.info("Assigning issue {} to user {}", issueKey, username);
        Map<String, String> request = Map.of("name", username);

        jiraRestClient.put()
                .uri("/issue/{issueKey}/assignee", issueKey)
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .toBodilessEntity();
    }

    /**
     * Get issues by project key
     */
    public JiraSearchResponse getIssuesByProject(String projectKey, int startAt, int maxResults) {
        log.info("Fetching issues for project: {}", projectKey);
        String jql = "project = " + projectKey + " ORDER BY created DESC";
        return searchIssues(jql, startAt, maxResults);
    }

    /**
     * Get issues assigned to a user
     */
    public JiraSearchResponse getAssignedIssues(String username, int startAt, int maxResults) {
        log.info("Fetching issues assigned to user: {}", username);
        String jql = "assignee = " + username + " AND status != Done ORDER BY priority DESC";
        return searchIssues(jql, startAt, maxResults);
    }
}

File: promptdev-frontend/lib/services/jira-service.ts
/**
 * Jira Server REST API service.
 * Direct HTTP calls to Jira Server.
 */

function getJiraConfig() {
  const baseUrl = process.env.JIRA_URL;
  const username = process.env.JIRA_USERNAME;
  const token = process.env.JIRA_TOKEN;
  const projectKey = process.env.JIRA_PROJECT_KEY;
  return { baseUrl, username, token, projectKey };
}

function getAuthHeaders(): Record<string, string> {
  const { username, token } = getJiraConfig();
  if (!username || !token) return {};
  const credentials = Buffer.from(`${username}:${token}`).toString("base64");
  return { Authorization: `Basic ${credentials}` };
}

export function isJiraConfigured(): boolean {
  const { baseUrl } = getJiraConfig();
  return !!baseUrl;
}

async function jiraApiFetch<T>(path: string, options?: RequestInit): Promise<T> {
  const { baseUrl } = getJiraConfig();
  if (!baseUrl) throw new Error("JIRA_URL is not configured");

  const url = `${baseUrl}/rest/api/2${path}`;
  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...getAuthHeaders(),
      ...options?.headers,
    },
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Jira API error ${response.status}: ${text}`);
  }

  const text = await response.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
}

export interface JiraSearchResult {
  issues: JiraIssueResult[];
  startAt: number;
  maxResults: number;
  total: number;
}

export interface JiraIssueResult {
  id: string;
  key: string;
  self: string;
  fields: {
    summary: string;
    description?: string;
    status: { name: string; id: string };
    assignee?: { name: string; displayName: string; emailAddress?: string };
    priority?: { name: string; id: string };
    issuetype: { name: string; id: string };
    project: { key: string; name: string };
    created: string;
    updated: string;
    labels?: string[];
    reporter?: { name: string; displayName: string };
  };
}

export interface JiraTransitionResult {
  id: string;
  name: string;
  to: { name: string; id: string };
}

export async function searchIssues(
  jql: string,
  startAt = 0,
  maxResults = 50,
): Promise<JiraSearchResult> {
  return jiraApiFetch<JiraSearchResult>(
    `/search?jql=${encodeURIComponent(jql)}&startAt=${startAt}&maxResults=${maxResults}`,
  );
}

export async function getIssue(issueKey: string): Promise<JiraIssueResult> {
  return jiraApiFetch<JiraIssueResult>(`/issue/${encodeURIComponent(issueKey)}`);
}

export async function getTransitions(
  issueKey: string,
): Promise<{ transitions: JiraTransitionResult[] }> {
  return jiraApiFetch<{ transitions: JiraTransitionResult[] }>(
    `/issue/${encodeURIComponent(issueKey)}/transitions`,
  );
}

export async function transitionIssue(
  issueKey: string,
  transitionId: string,
): Promise<void> {
  await jiraApiFetch<void>(`/issue/${encodeURIComponent(issueKey)}/transitions`, {
    method: "POST",
    body: JSON.stringify({ transition: { id: transitionId } }),
  });
}

export async function addComment(issueKey: string, comment: string): Promise<void> {
  await jiraApiFetch<void>(`/issue/${encodeURIComponent(issueKey)}/comment`, {
    method: "POST",
    body: JSON.stringify({ body: comment }),
  });
}

export async function assignIssue(issueKey: string, username: string): Promise<void> {
  await jiraApiFetch<void>(`/issue/${encodeURIComponent(issueKey)}/assignee`, {
    method: "PUT",
    body: JSON.stringify({ name: username }),
  });
}

export async function getIssuesByProject(
  projectKey: string,
  startAt = 0,
  maxResults = 50,
): Promise<JiraSearchResult> {
  const jql = `project = ${projectKey} ORDER BY created DESC`;
  return searchIssues(jql, startAt, maxResults);
}

export async function getAssignedIssues(
  username: string,
  startAt = 0,
  maxResults = 50,
): Promise<JiraSearchResult> {
  const jql = `assignee = ${username} AND status != Done ORDER BY priority DESC`;
  return searchIssues(jql, startAt, maxResults);
}
